### PR TITLE
[6X] add movetoMutex to protect moveto* fields in PGPROC

### PR DESCRIPTION
**Pull request by SmartKeyerror**
Actually, this PR is part of #13481, cause I consider that PR is too complicated
to merge directly, I decided to divide them and keep the core fix.

This PR is quite simple, just added a SpinLock to protect the field of `movetoResSlot`
and `movetoGroupId`, nothing else to say.

----

Since the race condition is not easy to add a test case to trigger it every time. If we
add inject_fault every time we find a race condition, the codes will be bulky.

### CODE REVIEW

File: src/backend/storage/ipc/procarray.c [hunk 1]
```diff
@@ -4945,11 +4945,15 @@ ResGroupSignalMoveQuery(int sessionId, void *slot, Oid groupId)
 
 		pid = proc->pid;
 		backendId = proc->backendId;
+
+		SpinLockAcquire(&proc->movetoMutex);
```
Comment (interma) [inline src/backend/storage/ipc/procarray.c:4949]: it's better to move it to in front of `Assert(proc->movetoResSlot == NULL);` (line 4953) and `Assert(groupId != InvalidOid);` (line 4962) ?
then the acquire and release is paired.

**Review by interma — APPROVED** [review]
LGTM, just small suggestion

File: src/backend/storage/ipc/procarray.c (hunk 1 above)
Comment (SmartKeyerror) [reply src/backend/storage/ipc/procarray.c]: Thanks for the great suggestion, already revise it.

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -4717,8 +4719,10 @@ HandleMoveResourceGroup(void)
 	}
 	else if (Gp_role == GP_ROLE_EXECUTE)
 	{
+		SpinLockAcquire(&MyProc->movetoMutex);
 		Oid groupId = MyProc->movetoGroupId;
 		MyProc->movetoGroupId = InvalidOid;
```
Comment (SmartKeyerror) [inline src/backend/utils/resgroup/resgroup.c:4724]: If there have 2 processes execute here and process A set `movetoGroupId` to `InvalidOid` and process B get `movetoGroupId`, it'll get an Invalid Oid, which will cause an error.

**Pull request merged**